solarforecastarbiter/validation/validator.py

```diff
@@ -406,3 +407,100 @@ def check_timestamp_spacing(times, freq=None):
         return len(gaps) == 1
     else:
         return True  # singleton DatetimeIndex passes
+
+
+def _zero_runs(a):
+    """ Finds runs of zeros in a.
+
+    Parameters
+    ----------
+    a : array
+
+
+    Returns
+    -------
+    ranges : array
+        Size Nx2, where ranges[k, 0] is the starting position of the kth run,
+        and ranges[k, 1] is the ending position, i.e., a[start:end] is zero.
+
+    Credit to https://stackoverflow.com/a/24892274/8484669
+    """
+    iszero = np.concatenate(([0], np.equal(a, 0).view(np.int8), [0]))
+    absdiff = np.abs(np.diff(iszero))
+    # Runs start and end where absdiff is 1.
+    ranges = np.where(absdiff == 1)[0].reshape(-1, 2)
+    return ranges
+
+
+def detect_stale_values(x, window=3, rtol=1e-5, atol=0):
+    """ Detects stale data.
+
+    Data are stale if the relative change in values is less than rtol within
+    window.
+
+    Parameters
+    ----------
+    x : Series
+        data to be processed
+    window : int, default 3
+        number of consecutive values which, if unchanged, are determined to be
+        stale data.
+    rtol : float, default 1e-5
+        relative tolerance for detecting a change in data values
+    atol : float, default 0
+        absolute tolerance for detecting a change in data values
+
+    Returns
+    -------
+    flags : Series
+        True if the value is part of a stale sequence of data
+    """
+
+    # identify pairs of elements that are nearly equal
+    close_to_next = np.isclose(x.values[1:], x.values[:-1], rtol=rtol,
+                               atol=atol)
+    runs = _zero_runs(np.logical_not(close_to_next))
```

Comment:
could you accomplish the same thing with something like:

```python
runs = pd.Series(close_to_next).rolling(window=window)
flags = runs.sum() == window
```

it looks similar with only a few minutes of testing.

Reply:
Didn't occur to me, I'll look into that.

Reply:
It's different behavior. I think the difference conceptually is that `.rolling.sum()` assigns a value to each window's right endpoint, whereas the _zero_runs function returns the two endpoints. I can probably figure out a way to do the calculation with `pd.Series.rolling` if you have a strong preference for that. I'll have to write additional lines to handle the ends of the Series where the window extends beyond the data.

Reply:
Ah, ok. I'd rather do whatever is simple (assuming performance is not a bottleneck). 

I'm not sure that `flags[r[0]:r[1]+1] = True` is what we want, though. Do we want to consider the first data points of the window to be stale? I wonder if the rolling behavior is what we actually want. 

```
data = [1.0, 1.001, 1.001, 1.001]
x = pd.Series(data=data)
res1 = validator.detect_interpolation(x, window=2)
# what should res1 = ?
[False, True, True, True]  # current behavior (I think)
[False, False, True, True]  # rolling behavior?
```

Reply:
That's a fair point - the first value in a sequence of equal values shouldn't be considered stale, but the ones following it should.  That behavior is easy to implement using `Series.rolling`.

Reply:
Pushed changes that use pd.Series.rolling instead of that _zero_run function.  Also, `window` now defines the number of consecutive values that have to meet the condition (e.g., stale) before a point is labeled.